clock_control: Fixes in the nrf driver

nordic-krch opened the pull request:
Start function had a race condition: clock was started and then user data was added to the list. It may happen that clock got started before item was added to the list due to short startup time or preemption.

When clock was started items were removed from the queue without protection. If it would collide with new clock start request, list would be corrupted. Refactored list handling, extracted it to helper functions.

nordic-krch committed:
drivers: clock_control: nrf: Fix race condition in clock start

If clock started immediately after requesting then request was
not yet placed in the list and user callback was not called.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nordic-krch committed:
drivers: clock_control: nrf: Add protection to getting item from list

When items were removed from the list there was not lock which may
lead to race condition and list corruption. Extracted list handling
into external function.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -282,16 +322,15 @@ static void clkstarted_handle(struct device *dev)
-	while (node != NULL) {
-		async_data = CONTAINER_OF(node,
-				struct clock_control_async_data, node);
+	while (async_data != NULL) {
```
What about:
```
	for (;;) {
		struct clock_control_async_data *async_data;
		
		async_data = list_get(&data->list);
		if (!async_data) {
			break;
		}

		async_data->cb(dev, async_data->user_data);
	}
```
The generated code should be simpler then, as `list_get` is called only once.
Or in a shorter version:
```
	while ((async_data = list_get(&data->list) != NULL) {
		async_data->cb(dev, async_data->user_data);
	}
```

nordic-krch replied:
done

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -149,34 +200,37 @@ static int clock_async_start(struct device *dev,
-			sys_slist_append(&clk_data->list, &data->node);
+		clock_irqs_disable();
+		started = clk_data->started;
+		if (!clk_data->started) {
```
Use of `started` (or `already_started` as suggested above) here would be a bit more appropriate, to indicate that this is a counterpart of the below code that calls `data->cb` directly.

nordic-krch replied:
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -140,6 +190,7 @@ static int clock_async_start(struct device *dev,
 						dev->config->config_info;
 	struct nrf_clock_control *clk_data = dev->driver_data;
 	int key;
+	bool started;
```
done

eobalski reviewed (approved):
Looks good.
If there is official bug reported please link to the PR.

pabigot commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -149,34 +199,47 @@ static int clock_async_start(struct device *dev,
-				return -EALREADY;
-			}
+		if (is_in_list(&clk_data->list, &data->node)) {
+			return -EALREADY;
```
At this point we've already incremented the reference count for this start operation, but we're returning an error rather than doing anything.  Is that correct management of the reference count?

pabigot reviewed (changes requested):
One comment below, but looking at this I have serious concerns about the correctness of the mutex management of the callback list, clock state, and ref count in this module.  Please give me a day or two to look more closely to be sure I understand what it's doing.  I'll have feedback NLT Monday morning EU time.

pabigot committed:
drivers: clock_control: nrf_power_clock: use unsigned reference count

The implementation did not account for the incrementing the reference
count past the maximum positive value.  Doing so on a signed integer
results in undefined behavior.  Switch to an unsigned integer where
such an increment results in a zero count, and add an assert that this
has not happened.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot reviewed (changes requested):
I'm proposing two changes to this:
* Move the check for error return before the ref count increment, which fixes the problem I identified in my first review.  This is blocking.
* Change the representation of the reference count to an unsigned value to avoid undefined behavior.  That's blocking too.

Patches for this are provided in https://github.com/pabigot/zephyr/commits/pr/20796.  You're welcome to cherry-pick those, or just squash them into your commits.

That said, I strongly believe that for correctness both `ref` and `list` must be maintained under the same mutex without releasing it between each operation.  This requires a significant refactoring, and when I [prototyped it](https://github.com/pabigot/zephyr/commits/refactor/20796) I became worried that the way stop and stop are handled is fragile, particularly around cases where a start is requested before a stop completes, and vice versa.  This also revealed some gaps in the API functionality and behavior.

I will approve this as a fix suitable for 2.1 with the blocking changes listed above, but in 2.2 asynchronous clock control should be revisited from both the API and implementation perspectives.  This should be coordinated with #20708, #20710, #20712, and #20806.

nordic-krch commented:
@pabigot I've applied your changes
>Move the check for error return before the ref count increment, which fixes the problem I identified in my first review. This is blocking.

Merged with my commit

Change the representation of the reference count to an unsigned value to avoid undefined behavior. That's blocking too.

cherry-picked

anangl commented:
> > Move the check for error return before the ref count increment, which fixes the problem I identified in my first review. This is blocking.
> 
> Merged with my commit

I think that in this case cherry-picking would also be a better choice. Currently, the commit message does not mention @pabigot's corrections at all, and these are quite significant, especially the change of returned error code.

pabigot commented:
I've approved this; I was waiting for a response to @anangl's concerns about the way mine were addressed.  I don't think there's particular need to highlight my changes, even the return value (which is not an API change but a bug fix to make the implementation match the API).

Dunno if it can make 2.1 at this point, but I don't object to it being a candidate.
